### PR TITLE
Feature flag & rollout parsing

### PR DESCRIPTION
**Pull request by AbdurRafay**
* New Models added

* variation variable usage mapping

* Feature variable and its usage parsing

* demoTestDatafile version upgraded to 4 and maintain compatibility to v3

* Test cases added for new models

* get back to Asynchronous Initialization of OPTLYManager

* CR updated

### CODE REVIEW

**Comment by kellyroach-optimizely** [comment]
32 XCODE  ANALYZE WARNINGS DETECTED

Xcode Clean / Build / Analyze the OptimizelySDKiOS / "iPhone X" .

The "master" branch currently generates ZERO compiler warnings (green check mark,
"No Issues").  However, P.R. #205 branch "oakbani:arafay/feature-flag-&-rollout-parsing"
is currently generating 32 Xcode Analyze warnings, such as:
```
../OptimizelySDKCore/OptimizelySDKCore/OPTLYVariation.h:36:40: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
@property (nonatomic, strong) NSString *variationId;
                                       ^
...
<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKCore' does not include header 'OPTLYFeatureFlag.h' [-Wincomplete-umbrella]
#import "Headers/OptimizelySDKCore.h"
^
<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKCore' does not include header 'OPTLYFeatureVariable.h' [-Wincomplete-umbrella]
<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKCore' does not include header 'OPTLYRollout.h' [-Wincomplete-umbrella]
<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKCore' does not include header 'OPTLYVariableUsage.h' [-Wincomplete-umbrella]
```

We ought to keep Xcode Clean / Build / Analyze the OptimizelySDKiOS / "iPhone X" 
at ZERO Xcode Analyze warnings if possible.  Can you take a look and fix these warnings?

**Review by kellyroach-optimizely — CHANGES_REQUESTED** [review]
32 new Xcode Analyze warnings have been added to P.R. #205, whereas "master" branch was at ZERO Xcode Analyze warnings.  Please fix if possible.

**Comment by kellyroach-optimizely** [comment]
P.R. #205 is for OASIS-1847 [Objective-C] Feature variable APIs

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYVariation.m [hunk 1]
```diff
@@ -16,14 +16,47 @@
...
+}
+
+- (NSDictionary<NSString *, OPTLYVariableUsage *> *)generateVariableIdToVariableUsageMap {
+    NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
+    for (OPTLYVariableUsage *variableUsage in self.variableUsageInstances) {
+        map[variableUsage.variableId] = variableUsage;
+    }
+    
+    return [NSDictionary dictionaryWithDictionary:map];
+}
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYVariation.m:60]: MISSING UNIT TESTS FOR THESE 3 NEW METHODS?

Are we missing Xcode unit tests for the three new methods
```
- (nullable OPTLYVariableUsage *)getVariableUsageForVariableId:(nullable NSString *)variableId {...}
- (NSDictionary<NSString *, OPTLYVariableUsage *> *)variableIdToVariableUsageMap {...}
- (NSDictionary<NSString *, OPTLYVariableUsage *> *)generateVariableIdToVariableUsageMap {...}
```
?

I ran OptimizelySDKCoreiOSTests with "Code Coverage" turned on.  These 3 new methods
weren't called in the "Show the Report navigator" "Coverage" report for OPTLYVariation.m
which is now only at 22.22% coverage.

**Comment by kellyroach-optimizely** [comment]
PLEASE USE Xcode'S "Sort by Name" WHEN ADDING NEW FILES

You've added new *.m+*.h files to the Xcode Project Navigator group
"OptimizelySDKCore/OptimizelySDKCore/Data Models" .  Can you "right click"
the "Data Models" tree element in the Xcode Project Navigator and select
"Sort by Name" from the popup context menu?  It's much easier to discover
files within a group when they are sorted alphabetically.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYVariation.m (hunk 1 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYVariation.m:60]: Yes. These 3 new methods will be covered in "getFeatureVariableValueForType" task where variable usage will be tested.
For now in order to increase code coverage I added basic test cases for these in OPTLYProjectConfigTest class under "Test [OPTLYVariation getVariableUsageForVariableId]:" mark. Now it is covered 100%

**Comment by kellyroach-optimizely** [comment]
CHECKING AGAIN AFTER commit eb8f415e9aa15be81a06108c98e8625b1ec8c5cb
"Removed xcode analyze warnings & code covered for OPTLYVariation.h"

TEST: OptimizelySDKiOS / "iPhone X" / Analyze
does not disappoint. ==> green check mark, "No issues".
PASSED
* iOS & tvOS demo apps PASSED
* OptimizelySDKTVOS / "Apple TV 4K" / Analyze PASSED
* 8+1 iOS, 7+1 tvOS test suites PASSED
* sh ./Scripts/build_all.sh PASSED

TEST: test code coverage for OPTLYVariation.m
* There are 2 new tests in the latest commit.
testGetVariableUsageForVariableId
testGetVariableUsageForVariableIdInvalid
* OPTLYVariation.m with the 3 new methods now has 100% test coverage.
PASSED

CONCLUDE: APPROVE

**Review by kellyroach-optimizely — APPROVED** [review]

**Pull request merged**